### PR TITLE
feat(downloads) - Adding section for `Video content download`

### PR DESCRIPTION
**Pull request by alexzchut**
## Description
Adding section for `Video content download`
## Checklist

* [x] PR is scoped to one task
* [ ] Tests are included
* [x] Documentation is included

* [x] This PR is not making any UI change
* [ ] This PR is making a UI change
  * [ ] Screenshot(s) included

## QA :

* required test cases:
* specific apps / plugins to test :

### CODE REVIEW

**Comment by ranmeirman** [comment]
@alexzchut do we have docs of how to a plug-in can declare itself as offline support?

**Comment by liviur** [comment]
just for clarification - this doc refers to adding a download button to an item in a screen / inside a cell referring to a specific item yeah?
This doesn't talk about enabling a plugin / screen to operate offline right?
What would we need to do to add that ability? ( i know the second part is what Ran refers to)

**Comment by alexzchut** [comment]
@ranmeirman Plugin does not declare itself for offline support, it is handling the download process and storage. The presentation of downloaded content is being made by the dedicated screen in UIB with DOWNLOADS datasource.

**Comment by alexzchut** [comment]
@liviur: 1. add the plugin. 2. add downloads screen that will display downloaded content.

**Comment by ranmeirman** [comment]
@alexzchut if I develop a screen plugin and I want it to work in offline mode, what should I do? Regardless of downloads.. imagine I want to create a screen and when the app is offline I want to show animation in the screen, what should I do? My question is not referring to downloads

**Comment by alexzchut** [comment]
@ranmeirman this documentation is not related to offline mode but to the content download only

**Comment by ranmeirman** [comment]
@alexzchut I understand but I am asking if we have this one, and if not we should add it :)

**Comment by alexzchut** [comment]
@ranmeirman we dont have

File: book/plugins/general-abilities/video-content-download/ios/video-content-download-ios.md [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+2. <a href="#general">General</a>
+3. <a href="#implementation">Download button protocol implementation</a>
+4. <a href="#download-button-states-images3
+">Download button states images</a>
+* * *
+
+<a name="description" />
+
+##### Description
+`Video Content download` ability uses the provided atom entry video url to download and save the content locally for offline playback.
```
Comment (liviur) [inline book/plugins/general-abilities/video-content-download/ios/video-content-download-ios.md]: Please add also a more specific designation of the guide.
`Given a screen plugin with Atom Item or items, the following guide goes through how to implement and customize a download button as a part of a Screen, component or cell.`

File: book/plugins/general-abilities/video-content-download/ios/video-content-download-ios.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+***
+
+<a name="general" />
+
+##### General
+
+Downloading can be performed for 2 types of videos `m3u8`, `mp4`
```
Comment (liviur) [inline book/plugins/general-abilities/video-content-download/ios/video-content-download-ios.md]:
```
Zapp apps supports downloads of one of the following video container types:
* HLS using m3u8 manifests - `m3u8` video type
* Mpeg video files for progressive download - `mp4` video type
```

File: book/plugins/general-abilities/video-content-download/ios/video-content-download-ios.md [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+- being made by implementing the downloading mechanism introduced in iOS 10 (and improved in iOS 11)
```
Comment (liviur) [inline book/plugins/general-abilities/video-content-download/ios/video-content-download-ios.md]: The offline video caching is implemented using Apple's built in video caching capabilities introduces in iOS 10.

File: book/plugins/general-abilities/video-content-download/ios/video-content-download-ios.md [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+- downloaded item (including its model, images, video) stored on device and can be presented on dedicated screen, played locally, or removed.
```
Comment (liviur) [inline book/plugins/general-abilities/video-content-download/ios/video-content-download-ios.md]: The Downloaded item (including the relevant metadata such as the Atom model and images) is stored on the device and will be used for presenting in a downloaded content dedicated screen, played locally, and marking it for deletion.

File: book/plugins/general-abilities/video-content-download/ios/video-content-download-ios.md (hunk 1 above)
Comment (liviur) [reply book/plugins/general-abilities/video-content-download/ios/video-content-download-ios.md]: designation - I mean scope ...

File: book/plugins/general-abilities/video-content-download/ios/video-content-download-ios.md [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+***
+
+<a name="protocol-implementation" />
```
Comment (liviur) [inline book/plugins/general-abilities/video-content-download/ios/video-content-download-ios.md]: please note - this doesn't work great in gitbook.
I would consider dropping it (test it locally to see)

File: book/plugins/general-abilities/video-content-download/ios/video-content-download-ios.md [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+##### Pre-condition
+Downloads plugin should be added to the app.
+
+##### Protocol implementation
+1. Implement `ZPDownloadButtonDelegate` to prepare the download button appearance on your screen.
```
Comment (liviur) [inline book/plugins/general-abilities/video-content-download/ios/video-content-download-ios.md]: this should be implemented on a button in the screen?
on the screen itself and then configure a button through the VC code?
Please elaborate

File: book/plugins/general-abilities/video-content-download/ios/video-content-download-ios.md [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+	- `func downloadButton(_ downloadButton: ZPDownloadButtonProtocol, stateChanged state: ZPDownloadButtonState)`
+		- fired on download button state change, usually no action is required on this function.
```
Comment (liviur) [inline book/plugins/general-abilities/video-content-download/ios/video-content-download-ios.md]: let's use `called` for the terminology. Also - please phrase it (and capitalize it) as a full sentence
Also - please add each function in a segment with 3 ` in order to keep it uniform.
Look at the login / analytics / player guide for expected structure.

Comment (liviur) [reply book/plugins/general-abilities/video-content-download/ios/video-content-download-ios.md]: This goes for the rest of the functions

File: book/plugins/general-abilities/video-content-download/ios/video-content-download-ios.md [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+		There are base sdk images for download button different states (explained in images section), custom suffix will be added to the image name lookup.
+		example:
+			 - item not downloaded image is `generic_item_download_btn` (value with custom suffix XYZ `generic_item_download_btnXYZ`)
+			 - item download progress image `generic_item_downloading_btn_1` (value with custom suffix XYZ `generic_item_downloading_btnXYZ_1`)
+
+2. Create a download button container on screen.
+
+3. Add download button to the container
+
+	- implementation example:
```
Comment (liviur) [inline book/plugins/general-abilities/video-content-download/ios/video-content-download-ios.md]: same - align it to the left please

File: book/plugins/general-abilities/video-content-download/ios/video-content-download-ios.md [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+
+	```swift
```
Comment (liviur) [inline book/plugins/general-abilities/video-content-download/ios/video-content-download-ios.md]: can you clean up this sample from "fluff" to minimal real world necessities ?
There's really no need for the constraints and stuff like that.

File: book/plugins/general-abilities/video-content-download/ios/video-content-download-ios.md [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+	```
+***
+
+<a name="images" />
+
+##### Download button states images
+
+Download button states default images are defined in Zapp SDK and attached to the app.
+- Mandatory button states images:
+	- Start Download Button image - `generic_item_download_btn`
```
Comment (liviur) [inline book/plugins/general-abilities/video-content-download/ios/video-content-download-ios.md]: For bullets we use `*` please keep to this style

File: book/plugins/general-abilities/video-content-download/ios/video-content-download-ios.md [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+	- Completed Download Button image - `generic_item_downloaded_btn`
+	- Error Download Button image - `generic_item_error_btn`
+
+- Optional button states images:
```
Comment (liviur) [inline book/plugins/general-abilities/video-content-download/ios/video-content-download-ios.md]: wrap optional like this `*Optional*`  - will turn it to italics

File: book/plugins/general-abilities/video-content-download/video-content-download-general.md [hunk 12]
```diff
@@ -0,0 +1,15 @@
+# Video Content Download
+
+`Video Content download` ability uses the provided atom entry video url to download and save the content locally for offline playback.
```
Comment (liviur) [inline book/plugins/general-abilities/video-content-download/video-content-download-general.md]: How much of this is duplicate from the iOS (or even worse - written twice with different styles)?

File: book/SUMMARY.md [hunk 13]
```diff
@@ -14,6 +14,8 @@
     - [Screen Hook](/plugins/general-abilities/screen-hooks/screen-hooks-general.md)
       - [Android](/plugins/general-abilities/screen-hooks/android/screen-hook-android.md)
       - [iOS](/plugins/general-abilities/screen-hooks/ios/screen-hook-ios.md)
+    - [Video Content Download](/plugins/general-abilities/video-content-download/video-content-download-general.md)
```
Comment (liviur) [inline book/SUMMARY.md]: Please note the updated plugin abilities location in the summary.md file

**Comment by liviur** [comment]
@alexzchut - added comments.
The offline mode support plugin capability will be a different PR I assume right?

File: book/plugins/general-abilities/video-content-download/ios/video-content-download-ios.md (hunk 5 above)
Comment (alexzchut) [reply book/plugins/general-abilities/video-content-download/ios/video-content-download-ios.md]: locally it works

File: book/SUMMARY.md (hunk 13 above)
Comment (alexzchut) [reply book/SUMMARY.md]: what do you mean, I added there..

Comment (alexzchut) [reply book/SUMMARY.md]: ok, added to the new place

**Comment by alexzchut** [comment]
@liviur - done

**Comment by ranmeirman** [comment]
@alexzchut, please add as a first section the direct API calls, and please remove the API to add a button, as the implementation is not completed and adding code to our scope (as discussed yesterday)

**Comment by alexzchut** [comment]
@ranmeirman  updating...

**Comment by alexzchut** [comment]
Added API calls

**Pull request merged**